### DISCUSSION TITLE
How do I use `Router::or` with `TraceLayer`?

### DISCUSSION THREAD

**Discussion by RabidFire** [tokio-rs/axum#335] [Q&A] [answered]
Hello! I'm trying to use `Router::or` but running into an issue. I want to separate my `/health` check route so that it doesn't get the TraceLayer applied to it.

    let api_router = Router::new()
        .route("/graphql", post(graphql_handler))
        .layer(TraceLayer::new_for_http());

    let health_router = Router::new().route("/health", get(health_handler));

    let app = api_router.or(health_router);

    axum_server::bind("0.0.0.0:8080").serve(app).await.unwrap()

But I get this weird error:

```
|     axum_server::bind("0.0.0.0:8080").serve(app).await.unwrap()
|                                       ^^^^^ expected struct `ResponseBody`, found struct `http_body::combinators::box_body::BoxBody`
```

Thought I followed the documentation pretty closely, but I'm not able to get past this. Works if I remove the TraceLayer. Works if I use either api_router or health_router separately. Would appreciate any help!

**Comment by davidpdrsn** [comment, marked as answer]
The issue here is that [`Trace`](https://docs.rs/tower-http/0.1.1/tower_http/trace/struct.Trace.html#associatedtype.Response) (applied by [`TraceLayer`](https://docs.rs/tower-http/0.1.1/tower_http/trace/struct.TraceLayer.html)) changes the response body type. By default axum's response body type is `axum::body::BoxBody` but [`Trace`](https://docs.rs/tower-http/0.1.1/tower_http/trace/struct.Trace.html#associatedtype.Response) changes that to [`tower_http::trace::ResponseBody`](https://docs.rs/tower-http/0.1.1/tower_http/trace/struct.ResponseBody.html). It does this because it needs to know when the body as been fully sent. You can see that in its [`Service` implementation](https://docs.rs/tower-http/0.1.1/tower_http/trace/struct.Trace.html#associatedtype.Response).

So the two routers don't have the same response body type meaning they cannot be combined. The solution is to call `.boxed()` on the one with the `TraceLayer` so `.layer(TraceLayer::new_for_http()).boxed()`. That changes the response body type so they now both match.

Here is a full example:

```rust
use axum::{
    handler::get,
    Router,
};
use std::net::SocketAddr;
use tower_http::trace::TraceLayer;

#[tokio::main]
async fn main() {
    let api_router = Router::new()
        .route("/graphql", get(|| async {}))
        .layer(TraceLayer::new_for_http())
        .boxed();

    let health_router = Router::new().route("/health", get(|| async {}));

    let app = api_router.or(health_router);

    // run it
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    tracing::debug!("listening on {}", addr);
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}
```
